### DISCUSSION TITLE
Values other than string for Select

### DISCUSSION THREAD

**Discussion by 7iomka** [mantinedev/mantine#5086] [Q&A]
I have Select with number values, and onChange handler that need to send this numeric values.
In v6 it worked, maybe internally somehow transformed intro string.
In v7 I receive error
<img width="976" alt="image" src="https://github.com/mantinedev/mantine/assets/18473137/5f347928-60a8-4fdd-80ff-123cf4e9760a">
What was the purpose of throw error like this?

**Comment by rtivital** [comment]
Select option values were not converted to string, the purpose of the error is to emphasize that only string data format is supported. Otherwise, the component will throw in another place with an unclear message.

**Reply by Nikitakun** [reply]
Hi @rtivital, I have come across the same issue. Would appreciate your feedback. 

Typically in the majority of cases for enterprise applications, selects, autocompletes and components alike are passed numeric ids of elements and are expected to receive a numeric id back on selection.
Currently, such super frequent use case will have to be processed with back and forth mapping - for options, we gotta map a list of numeric ids to strings and pass them as `data`, then for controlled components we gotta convert the selected value to string to pass it to `value`, and `onChange` we gotta convert the selected string back to number. 

Additionally, it is often convenient to pass an object as a value and receive it back as well. I understand there is minimal support for this for single selects in onChange, however seemingly not for multi-selects or autocompletes. And there is no support for passing objects as options and values.

To my understanding, implementing it for `Combobox` and subsequently for all other components build on top of it would be  straightforward with simple === value comparisons where needed for determining selected values, and it would be easy to even add an optional comparator function as a prop for comparing potentially complex values if needed which would also be likely just a few lines of code.

My belief is that it's going to cover very common use cases and would not be difficult to implement, whereas currently it seems that we have to do lots of unnecessary mapping for a very simple use case.

**Reply by rtivital** [reply]
Select component will support primitive data types starting from version 9.0, it is already implemented and will be available later this year. Before that you will need to transform data passed to the component into strings.

**Reply by Nikitakun** [reply]
Amazing news, thank you! v9.0 should be expected in the coming couple of months?

**Reply by BrendanC23** [reply]
Version [9.0.0-alpha](https://alpha.mantine.dev/changelog/9-0-0/) is available. See [here](https://alpha.mantine.dev/changelog/9-0-0/#generic-components) for details on using non-string primitives with `Select`.

>SegmentedControl, Select, MultiSelect, Chip.Group, Switch.Group, Checkbox.Group and Radio.Group now support generic value type. You can pass primitive values (numbers, strings, boolean, null) as the type argument. The generic type is used for value, defaultValue, onChange and other props.

See [here](https://alpha.mantine.dev/core/select/#value-type) for an example with `Select`.

>Select supports primitive values (strings, numbers, booleans) as value type. Select automatically infers the value type. If you want to set the value type explicitly, pass type argument:

```tsx
type SelectValue = 'React' | 'Angular' | 'Svelte' | number;

function Demo() {
  return <Select<SelectValue> data={['React', 'Angular', 'Svelte', 100]} />;
}
```
